### DISCUSSION TITLE
Doctrine Parameter Converter injecting first collection member when no value is provided

### DISCUSSION THREAD

**Discussion by janklan** [symfony/symfony#47166] [Q&A] [answered]
I tripped over a peculiar behaviour involving the [SensioFrameworkExtraBundle's DoctrineParamConverter](https://github.com/sensiolabs/SensioFrameworkExtraBundle/blob/master/src/Request/ParamConverter/DoctrineParamConverter.php#L209), and now also [Entity Argument Resolver](https://github.com/symfony/symfony/pull/43854/files#diff-d7a94dd8064136291feb3ebaec47ada47c13f6eaf683abdd5be49a8d5e6de32eR200) in https://github.com/symfony/symfony/pull/43854.

In a common controller URL parameter conversion scenario, it goes like this: 

1. If you have a route: `#[Route('/test/{parentEntity<\d+>}/load/{childEntity<\d+>?})]`, and
2. If your controller has a signature of `public function action(Parent $parentEntity, ?Child $childEntity)`, then
3. When you call `/test/50/load/`, the `childEntity` should be `null`, but instead it is whatever is the first member of the child collection in the `parentEntity` is.

The root cause of this is the piece of code I linked above:

```php
if (null === $mapping = $options->mapping) {
    $keys = $request->attributes->keys();
    $mapping = $keys ? array_combine($keys, $keys) : [];
}

// ... 

$metadata = $objectManager->getClassMetadata($class); // Child entity metadata
foreach ($mapping as $attribute => $field) {
    if (!$metadata->hasField($field) && (!$metadata->hasAssociation($field) || !$metadata->isSingleValuedAssociation($field))) {
        continue;
    }

    $criteria[$field] = $request->attributes->get($attribute);
}

// ...

try {
    return $objectManager->getRepository($class)->findOneBy($criteria);
} catch (NoResultException|ConversionException) {
    return null;
}
```

1. The `$mapping` contains the current request attributes, so for instance, `['parentEntity' => 50, '_route' => '_route', ...]`. 
2. **Here is the problem:** If the `$child` happens to have an association called `parentEntity`, the code above slaps `['parentEntity' => 50]` to `$criteria`. 
3. `$criteria` is later passed to `findOneBy`. Since the `criteria` only contains the `parentEntity`'s ID, it will find one `childEntity` according to the `$criteria` - the first that belongs to the parent with ID 50. **But we did not pass any childEntity ID - we did not want the converter to load anything**

---

I think the original idea is not bad - the end effect is that if there are multiple related entities, they are guaranteed to belong together, because the `$criteria` contains more than just a single primary key. Even though it's magic that depends on consistent naming conventions, let's roll with it. 

The problem I see is that if the `childEntity` value is not present in the `$mapping` array (= the value is missing in the Request), none of this code should have been executed in the first place. 

This behaviour can be suppressed by [providing the `id` param name in options](https://github.com/symfony/symfony/pull/43854/files#diff-d7a94dd8064136291feb3ebaec47ada47c13f6eaf683abdd5be49a8d5e6de32eR192), but I'm wondering if this should not be the default for optional parameters?

**Comment by MatTheCat** [comment]
> The problem I see is that if the `childEntity` value is not present in the `$mapping` array (= the value is missing in the Request), none of this code should have been executed in the first place.

Only you can know this so I don’t think the default behavior will change!

> The converter will attempt to do a `findOneBy()` fetch by using all of the wildcards in your route that are actually properties on your entity.
>
> https://symfony.com/bundles/SensioFrameworkExtraBundle/current/annotations/converters.html#1-fetch-automatically

There are options like `id` to address use cases where automatic fetching does not work, do not be afraid to use them :ok_hand: 

---

Your route’s pattern is also weird because you can get the parent from the child so only one parameter would be needed?

**Reply by janklan** [reply]
> Only you can know this so I don’t think the default behavior will change!

It works well when both arguments are present, but the default is to load "something" (= first child of a parent) when the value is "nothing" (= null). 

> Your route’s pattern is also weird because you can get the parent from the child so only one parameter would be needed?

I have good reasons for doing so. The bottom line is that the current default seems to be not ideal. If the child argument is not provided, it should stay null.

**Reply by MatTheCat** [reply]
Okay I understand now (sorry I wasn’t aware a route parameter was considered an ID if its name matches an entity’s).

Indeed the fact `null` and `false` are treated the same in https://github.com/symfony/symfony/blob/1fa454375eaf975d613bf6804031bf0955c6316f/src/Symfony/Bridge/Doctrine/ArgumentResolver/EntityValueResolver.php#L125-L127 feels weird. I’ll check that out.

---

Sooo no tests are covering this behavior (well [`testApplyGuessOptional`](https://github.com/symfony/symfony/blob/082090adc1f97b5d1cd098ef6981abd628da6a97/src/Symfony/Bridge/Doctrine/Tests/ArgumentResolver/EntityValueResolverTest.php#L273-L304) would complain, but it seems unrelated).

I opened https://github.com/symfony/symfony/pull/47242

**Reply by nicolas-grekas** [reply]
> If the child argument is not provided, it should stay null.

Could be worth considering. Would you mind sending a PR that does so?

**Reply by MatTheCat** [reply]
@nicolas-grekas I opened https://github.com/symfony/symfony/pull/47242 but GitHub won't link discussions 😔

**Comment by nicolas-grekas** [comment, marked as answer]
Fixed by https://github.com/symfony/symfony/pull/47242 from @MatTheCat :pray:

**Comment by janklan** [comment]
Thanks @MatTheCat 

FYI the real-world scenario for my weird routes: I have multi-tenant app where I prefix every URL with tenant's slug. Say there was Child entity owned by a Tenant. In my routes, I used `/{tenant}/{child<\d+>?}`. When I called /tenant-id entity, I wanted to show some Tenant-related page. When I called /tenant-id/child-id, I wanted to show the same page, but then include some additional info about that selected child. 

Since Child had ManyToOne relation to Tenant, the param converter always tried to find the first child when the ID was null. 

I'm using the Tenant entity in the URL for access control. One of my early event subscribers (kernel.controller with 4096 priority) figures out what organisation is the user accessing, checks permission to do so for the user in session, and activates row level security in PostgreSQL so that data of other tenants does no longer show up, and then lets the lest of the request go trhough, incl. other route parameter conversion etc. 

I decided I have the tenant present in every tenant-aware route instead of session, where it created some complexities that were not worth the hassle.
